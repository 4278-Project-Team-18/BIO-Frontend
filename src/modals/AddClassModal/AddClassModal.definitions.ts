/* eslint-disable autofix/no-unused-vars */
import type { Teacher } from "../../interfaces/User.interface";

export interface AddClassModalProps {
  closeModal: () => void;
  teacher?: Teacher;
}

/**
 * Interface for the AddClassModal component
 *
 * @param {string} className - name of the class
 * @param {string} teacherId - id of the teacher
 */
export interface NewClassInput {
  name: string;
  teacherId: string;
}

export enum NewClassInputName {
  CLASS_NAME = "name",
  TEACHER_ID = "teacherId",
}
