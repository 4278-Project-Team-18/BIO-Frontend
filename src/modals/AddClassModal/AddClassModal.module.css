.add-class-backdrop {
  position: fixed;
  z-index: 1;
  left: 0;
  top: 0;
  width: 100%;
  height: 100%;
  overflow: auto;
  background-color: rgb(0, 0, 0);
  background-color: rgba(0, 0, 0, 0.5);
}

.add-class-container {
  background-color: #fefefe;
  margin: 20% auto;
  padding: 20px;
  width: fit-content;
  min-width: 40%;
  border-radius: 20px;
}

.add-class-title {
  font-size: 20px;
  font-weight: 600;
  margin: 0;
  padding: 0;
  margin-left: 3px;
  color: #414141;
}

.add-class-top-section {
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
  padding-bottom: 10px;
}

.add-class-cancel-button {
  font-size: 20px;
  padding-left: 8px;
  padding-right: 10px;
  padding-bottom: 2px;
  color: #787878;
}
