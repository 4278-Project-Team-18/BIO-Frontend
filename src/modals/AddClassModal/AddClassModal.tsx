import { NewClassInputName } from "./AddClassModal.definitions";
import styles from "./AddClassModal.module.css";
import FormInput from "../../components/FormInput/FormInput";
import { newClassSchema } from "../../resolvers/class.resolver";
import FormSelect from "../../components/FormSelect/FormSelect";
import { RequestMethods, useCustomFetch } from "../../api/request.util";
import { useClassesContext } from "../../context/Classes.context";
import LoadingButton from "../../components/LoadingButton/LoadingButton";
import { useForm } from "react-hook-form";
import { yupResolver } from "@hookform/resolvers/yup";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faCircleXmark, faPlusCircle } from "@fortawesome/free-solid-svg-icons";
import { useEffect } from "react";
import type { FormSelectOption } from "../../components/FormSelect/FormSelect.definition";
import type { Class, Teacher } from "../../interfaces/User.interface";
import type {
  AddClassModalProps,
  NewClassInput,
} from "./AddClassModal.definitions";
import type { Resolver } from "react-hook-form";

const AddClassModal = ({ closeModal }: AddClassModalProps) => {
  // request to get all teachers
  const {
    data: teacherData,
    loading: teacherLoading,
    error: teacherError,
  } = useCustomFetch<Teacher[]>(`teacher/allTeachers`);

  const {
    data: classData,
    loading: classLoading,
    // error: classError,
    makeRequest: makeClassRequest,
  } = useCustomFetch<Class>(`class`, RequestMethods.POST);

  const { addClass } = useClassesContext();

  // form controller
  const {
    control,
    handleSubmit,
    formState: { errors },
    setValue,
    // setError,
  } = useForm<NewClassInput>({
    defaultValues: {
      [NewClassInputName.CLASS_NAME]: "",
      [NewClassInputName.TEACHER_ID]: "",
    },
    resolver: yupResolver(newClassSchema) as Resolver<NewClassInput>,
    mode: "onSubmit",
  });

  // send the request to add the class
  const onSubmitNewClass = async (inputData: NewClassInput) => {
    await makeClassRequest(inputData);
  };

  // set the default value of the teacher select to the first teacher
  useEffect(() => {
    if (teacherData && teacherData.length > 0) {
      setValue(NewClassInputName.TEACHER_ID, teacherData[0]._id);
    }
  }, [teacherData]);

  useEffect(() => {
    if (classData) {
      addClass(classData);
      closeModal();
    }
  }, [classData]);

  // if the request is loading, show a loading message
  if (teacherLoading) {
    return (
      <div className={styles["add-class-backdrop"]}>
        <div>Loading...</div>;
      </div>
    );
  }

  // if the request failed, show an error message
  if (teacherError || !teacherData) {
    return (
      <div className={styles["add-class-backdrop"]}>
        <div>Something went wrong...</div>
      </div>
    );
  }

  // map the teacher data to the form options
  const formOptions = teacherData.map(
    (teacher: Teacher) =>
      ({
        value: teacher._id,
        label: `${teacher.firstName} ${teacher.lastName}`,
      }) as FormSelectOption,
  );

  return (
    <div className={styles["add-class-backdrop"]}>
      <form
        className={styles["add-class-container"]}
        onSubmit={handleSubmit(onSubmitNewClass)}
      >
        <div className={styles["add-class-top-section"]}>
          <div className={styles["add-class-title"]}>Add Class</div>
          <button
            className={styles["add-class-cancel-button"]}
            onClick={closeModal}
            type="button"
            aria-label="close-button"
          >
            <FontAwesomeIcon
              icon={faCircleXmark}
              className={styles["add-class-cancel-button-icon"]}
            />
          </button>
        </div>
        <div>
          <FormInput
            control={control}
            type="text"
            name={NewClassInputName.CLASS_NAME}
            label="Class Name"
            error={errors[NewClassInputName.CLASS_NAME]?.message}
            placeholder="Class Name"
          />
          <FormSelect
            control={control}
            name={NewClassInputName.TEACHER_ID}
            label="Teacher"
            error={errors[NewClassInputName.TEACHER_ID]?.message}
            options={formOptions}
            defaultValue={formOptions[0].value}
            setValue={setValue}
          />

          <div className={styles["add-class-submit-container"]}>
            <div className={styles["add-class-submit-inner-container"]}>
              <LoadingButton
                text="Add Class"
                type="submit"
                icon={faPlusCircle}
                isLoading={classLoading}
                isLoadingText="Adding Class..."
              />
            </div>
          </div>
        </div>
      </form>
    </div>
  );
};

export default AddClassModal;
