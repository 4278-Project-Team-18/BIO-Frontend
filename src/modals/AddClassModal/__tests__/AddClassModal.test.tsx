import AddClassModal from "../AddClassModal";
import { render } from "@testing-library/react";
import { act } from "react-dom/test-utils";
import type { Teacher } from "../../../interfaces/User.interface";

// Mocking the useCustomFetch module
jest.mock("../../../api/request.util", () => ({
  useCustomFetch: () => ({
    data: [
      {
        _id: "12345",
        firstName: "John",
        lastName: "Doe",
        email: "testemail@test.com",
        role: "teacher",
      } as Teacher,
    ] as Teacher[],
    loading: false,
    error: null,
    makeRequest: jest.fn(),
  }),
  RequestMethods: {
    POST: "POST",
    GET: "GET",
  },
}));

describe("AddClassModal Tests", () => {
  test("AddClassModal should render correctly", () => {
    const mockProps = {
      closeModal: jest.fn(),
    };

    const component = render(<AddClassModal {...mockProps} />);
    expect(component).toBeTruthy();

    // expect(component.getByText("Add class")).toBeTruthy();
    expect(component.getByText("Class Name")).toBeTruthy();

    // Check if the buttons are rendered
    expect(component.getAllByText("Add Class")).toBeTruthy();

    // check for teacher in the select
    expect(component.getByText("John Doe")).toBeTruthy();

    // click on the add class button
    act(() => {
      component.getAllByText("Add Class")[0].click();
    });
    expect(mockProps.closeModal).toHaveBeenCalled();

    // click on the close button
    act(() => {
      component.getByLabelText("close-button").click();
    });
    expect(mockProps.closeModal).toHaveBeenCalled();
  });

  test("AddClassModal snapshot should render correctly", () => {
    const mockProps = {
      closeModal: jest.fn(),
    };

    const component = render(<AddClassModal {...mockProps} />);
    expect(component).toMatchSnapshot();
  });
});
