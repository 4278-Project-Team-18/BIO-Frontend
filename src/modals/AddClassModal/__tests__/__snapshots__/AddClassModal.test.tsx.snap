// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddClassModal Tests AddClassModal snapshot should render correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="add-class-backdrop"
      >
        <form
          class="add-class-container"
        >
          <div
            class="add-class-top-section"
          >
            <div
              class="add-class-title"
            >
              Add Class
            </div>
            <button
              aria-label="close-button"
              class="add-class-cancel-button"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-circle-xmark add-class-cancel-button-icon"
                data-icon="circle-xmark"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 512 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M256 512A256 256 0 1 0 256 0a256 256 0 1 0 0 512zM175 175c9.4-9.4 24.6-9.4 33.9 0l47 47 47-47c9.4-9.4 24.6-9.4 33.9 0s9.4 24.6 0 33.9l-47 47 47 47c9.4 9.4 9.4 24.6 0 33.9s-24.6 9.4-33.9 0l-47-47-47 47c-9.4 9.4-24.6 9.4-33.9 0s-9.4-24.6 0-33.9l47-47-47-47c-9.4-9.4-9.4-24.6 0-33.9z"
                  fill="currentColor"
                />
              </svg>
            </button>
          </div>
          <div>
            <div
              class="form-input-container"
            >
              <div
                class="form-input-inner-container"
              >
                <div
                  class="form-input-label-opt"
                >
                  Class Name
                </div>
                <input
                  class="form-input"
                  name="name"
                  placeholder="Class Name"
                  type="text"
                  value=""
                />
              </div>
            </div>
            <div
              class="form-select-container"
            >
              <div
                class="form-select-inner-container"
              >
                <div
                  class="form-select-label-opt"
                >
                  Teacher
                </div>
                <select
                  class="form-select"
                  name="teacherId"
                >
                  <option
                    class="form-select-option"
                    selected=""
                    value="12345"
                  >
                    John Doe
                  </option>
                </select>
              </div>
            </div>
            <div
              class="add-class-submit-container"
            >
              <div
                class="add-class-submit-inner-container"
              >
                <button
                  class="loading-button"
                  style="background-color: rgb(90, 208, 109);"
                  type="submit"
                >
                  <span
                    class="loading-button-label"
                  >
                    Add Class
                  </span>
                  <svg
                    aria-hidden="true"
                    class="svg-inline--fa fa-circle-plus loading-button-spinner"
                    data-icon="circle-plus"
                    data-prefix="fas"
                    focusable="false"
                    role="img"
                    viewBox="0 0 512 512"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M256 512A256 256 0 1 0 256 0a256 256 0 1 0 0 512zM232 344V280H168c-13.3 0-24-10.7-24-24s10.7-24 24-24h64V168c0-13.3 10.7-24 24-24s24 10.7 24 24v64h64c13.3 0 24 10.7 24 24s-10.7 24-24 24H280v64c0 13.3-10.7 24-24 24s-24-10.7-24-24z"
                      fill="currentColor"
                    />
                  </svg>
                </button>
              </div>
            </div>
          </div>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="add-class-backdrop"
    >
      <form
        class="add-class-container"
      >
        <div
          class="add-class-top-section"
        >
          <div
            class="add-class-title"
          >
            Add Class
          </div>
          <button
            aria-label="close-button"
            class="add-class-cancel-button"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-circle-xmark add-class-cancel-button-icon"
              data-icon="circle-xmark"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 512 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M256 512A256 256 0 1 0 256 0a256 256 0 1 0 0 512zM175 175c9.4-9.4 24.6-9.4 33.9 0l47 47 47-47c9.4-9.4 24.6-9.4 33.9 0s9.4 24.6 0 33.9l-47 47 47 47c9.4 9.4 9.4 24.6 0 33.9s-24.6 9.4-33.9 0l-47-47-47 47c-9.4 9.4-24.6 9.4-33.9 0s-9.4-24.6 0-33.9l47-47-47-47c-9.4-9.4-9.4-24.6 0-33.9z"
                fill="currentColor"
              />
            </svg>
          </button>
        </div>
        <div>
          <div
            class="form-input-container"
          >
            <div
              class="form-input-inner-container"
            >
              <div
                class="form-input-label-opt"
              >
                Class Name
              </div>
              <input
                class="form-input"
                name="name"
                placeholder="Class Name"
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            class="form-select-container"
          >
            <div
              class="form-select-inner-container"
            >
              <div
                class="form-select-label-opt"
              >
                Teacher
              </div>
              <select
                class="form-select"
                name="teacherId"
              >
                <option
                  class="form-select-option"
                  selected=""
                  value="12345"
                >
                  John Doe
                </option>
              </select>
            </div>
          </div>
          <div
            class="add-class-submit-container"
          >
            <div
              class="add-class-submit-inner-container"
            >
              <button
                class="loading-button"
                style="background-color: rgb(90, 208, 109);"
                type="submit"
              >
                <span
                  class="loading-button-label"
                >
                  Add Class
                </span>
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-circle-plus loading-button-spinner"
                  data-icon="circle-plus"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  viewBox="0 0 512 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M256 512A256 256 0 1 0 256 0a256 256 0 1 0 0 512zM232 344V280H168c-13.3 0-24-10.7-24-24s10.7-24 24-24h64V168c0-13.3 10.7-24 24-24s24 10.7 24 24v64h64c13.3 0 24 10.7 24 24s-10.7 24-24 24H280v64c0 13.3-10.7 24-24 24s-24-10.7-24-24z"
                    fill="currentColor"
                  />
                </svg>
              </button>
            </div>
          </div>
        </div>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
